docs: adding dynamic channel page

mhmohona opened the pull request:
**Description**
 Added **Dynamic channel name** page.

It is a part of GSoD'23 project.


**Related issue(s):** fixes ##1511

mhmohona committed:
adding dynamic channel page

smoya commented on pages/docs/concepts/asyncapi-document/dynamic_channel_names.md:
```diff
@@ -0,0 +1,117 @@
+        context: "payload"
+        schema:
+          type: "string"
+    subscribe:
```
This example is not valid in AsyncAPI v3. Please see https://github.com/asyncapi/website/pull/1868#discussion_r1246407135.

If you need to test your code examples, you can do it manually by following the steps from https://asyncapi.slack.com/archives/C0230UAM6R3/p1689352232934939

mhmohona replied:
Thank you for your review @smoya. 
Shall i update all code according to [this](https://asyncapi.slack.com/archives/C0230UAM6R3/p1689352232934939) format?

smoya replied:
This doc should be based on AsyncAPI v3, which is a WIP but the changes on the spec can be read at https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12

smoya commented on pages/docs/concepts/asyncapi-document/dynamic_channel_names.md:
```diff
@@ -0,0 +1,117 @@
+weight: 220
+---
+
+In AsyncAPI, dynamic parts of a channel name can be specified using parameters. These parameters are typically defined in the components section of AsyncAPI document. They can be used throughout the document and specifically, they can be referenced in the channels section to denote dynamic parts of channel names.
```
I would not emphasize the fact you can reuse parameters by adding them to the components. It's not important in this first paragraph when still the concept is not properly explained. In fact, you wrote an entire section for that: https://github.com/asyncapi/website/pull/1959/files#diff-3ce57f81bfffd1ad053da3175b59c0928e874a7cd30dec23b035662c07030b8bR93

I would additionally mention the reason I would need to use those[ channel address expressions](ttps://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#channelAddressExpressions) (the real name btw), showing examples, etc.

smoya commented on pages/docs/concepts/asyncapi-document/dynamic_channel_names.md:
````diff
@@ -0,0 +1,117 @@
+
+```yml
+channels:
+  user/{userId}/signup:
````
Channel address in AsyncAPI v3 is described in the `address` field of the channel.

smoya commented on pages/docs/concepts/asyncapi-document/dynamic_channel_names.md:
```diff
@@ -0,0 +1,117 @@
+      userId:
+        description: "Id of the user."
+        context: "payload"
+        schema:
```
Channel parameters shape in AsyncAPI v3 are different. They are not expected to be schemas anymore but simplified types. See https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#parameterObject

mhmohona committed:
update code

quetzalliwrites commented on pages/docs/concepts/asyncapi-document/dynamic_channel_names.md:
```diff
@@ -0,0 +1,125 @@
+title: Dynamic Channel Names
+weight: 220
+---
+
```
Please fix your file name to replace underscores with dashes: `dynamic-channel-names.md`.

quetzalliwrites replied:
Pretty well-written and informative! ✨✨ but could benefit from additional context and organization. ✌🏽

---

1. **Introduction**: The introduction could be expanded to give a bit more background on the importance and advantages of using dynamic channel names in AsyncAPI. Could you mention **why the feature is used** and **how it can benefit folks or systems**?

2. **Explanation**: Some parts of the document could use a more detailed explanation. For example, when introducing the concept of _Parameter Context_, you could explain why it is important and how it aids in the user experience?

3. **Organization**: You might want to reorganize the document a bit to improve the flow. For instance, before diving into 'Parameter Context', it might be better to introduce 'Reusing Parameters' since it directly follows the initial introduction of parameters in channels.

4. **Graphics**: Ensure your diagrams accurately represent the text as _clearly as possible_. The second diagram starting on line 57, for example, might be clearer if it included the word `payload`, to match what you wrote it in the text description introducing the diagram (i.e., "runtime event payload").

5. **Proofreading**: There are grammatical/spelling errors (like 'vesion' instead of 'version' in one of the YAML examples on line 81) that should be corrected. Please review this doc and fix those mistakes. 😄 

6. **Do not use dashes `-` instead of colons `:` at the end of a sentence.**

quetzalliwrites reviewed (changes requested):
time to work, Miss Mhafuzaaaa ✨✨✨✨✨ 😄✌🏽

mhmohona committed:
updating as per review

mhmohona committed:
update weight

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
```diff
@@ -0,0 +1,84 @@
+---
+title: Dynamic Channel Names
```
We talk about dynamic addresses. In previous versions, the address was also the name of the channel, but in v3, as I mentioned in https://github.com/asyncapi/website/pull/1959#discussion_r1265651267, it's moved into the `address` field.

This should be changed across the whole doc. 
```suggestion
title: Dynamic Channel Addresses
```

Or 

```suggestion
title: Dynamic Channel Address
```

Not sure what sounds better.

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
```diff
@@ -0,0 +1,84 @@
+weight: 80
+---
+
+Dynamic channel names, also known as channel address expressions, are a key feature of AsyncAPI. Dynamic channel names in AsyncAPI are pivotal for enhancing the flexibility, interoperability, reusability, and standardization of asynchronous APIs. They allow for variable usage in channel names that can be adapted to multiple APIs, used across different protocols, and reused across channels, thus p ...
```
nitpick: I had to search for the term `pivotal` 😅 Not sure if you could replace by a more known one. Anyway, discard my suggestion if the word is well used out there and I'm the only one with the issue 😆

quetzalliwrites replied:
sure, we can use a synonym that seems less jargony such as `essential` or other ✌🏽

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
```diff
@@ -0,0 +1,84 @@
+weight: 80
+---
+
+Dynamic channel names, also known as channel address expressions, are a key feature of AsyncAPI. Dynamic channel names in AsyncAPI are pivotal for enhancing the flexibility, interoperability, reusability, and standardization of asynchronous APIs. They allow for variable usage in channel names that can be adapted to multiple APIs, used across different protocols, and reused across channels, thus p ...
```
> Dynamic channel names, also known as channel address expressions, are a key feature of AsyncAPI. Dynamic channel names in AsyncAPI are pivotal for enhancing the flexibility, interoperability, reusability, and standardization of asynchronous APIs. They allow for variable usage in channel names that can be adapted to multiple APIs, used across different protocols, and reused across channels, thus preventing redundancy and maintaining consistency. This adaptability to different APIs aids in creating a unified view of the API, making it easier to comprehend and utilize. Dynamic channel names contribute to code and documentation generation and event management tooling, thereby improving APIs' overall efficiency and usability.

I like how this sentence sounds in overall, however I don't feel it provides much value to the user. Let me rephrase that. IMHO it adds too many good words that are not really important or related to the matter we talk about. I.e. there is only one (short) part of one sentence that explains the subject: ` They allow for variable usage in channel names`.
I would advocate for short, concise, and simpler documentation, without (much) ornaments. 

What do you think about focus on what it technically means (define dynamic values in the address AKA [Channel address expressions](https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#channelAddressExpressions)) and tell by naming real use cases? I believe this will help users to understand what are talking about here.

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
```diff
@@ -0,0 +1,84 @@
+
+## Parameter Context
+
+The Parameter Context clarifies the origin of parameters, ensuring consistent understanding across teams. It also enables efficient code generation, thereby accelerating development and enhancing the developer experience.
```
I would recommend to apply the same suggestion I wrote in https://github.com/asyncapi/website/pull/1959/files#r1281562860

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
```diff
@@ -0,0 +1,84 @@
+
+The Parameter Context clarifies the origin of parameters, ensuring consistent understanding across teams. It also enables efficient code generation, thereby accelerating development and enhancing the developer experience.
+
+In AsyncAPI, the context of parameters in channel names often refers to where the parameter value comes from. For example, the parameter might be populated from the runtime event payload. This context can be noted in the AsyncAPI document, and code generators can use it to make the developer experience easier.
```
> to make the developer experience easier.

WDYT about adding the "how" to this statement? I mean explaining "How makes it easier?"?

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
```diff
@@ -0,0 +1,84 @@
+
+## Reusing Parameters
+
+An important thing about parameters is that they can be reused. If there is another event, for example, a `UserUpdated` event, which also requires the userId, the parameter can be reused like the following example:
```
Applying our ubiquitous language, `message` is better than `event` here.

```suggestion
An important thing about parameters is that they can be reused. If there is another message, for example, a `UserUpdated` message, which also requires the userId, the parameter can be reused like the following example:
```

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
````diff
@@ -0,0 +1,84 @@
+Here is an example of parameter context:
+
+```yml
+address: user/{userId}/signedup
````
Nitpick: for consistency with examples in our reference docs:

```suggestion
address: 'user/{userId}/signedup'
```

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
````diff
@@ -0,0 +1,84 @@
+```yml
+channels:
+  userSignedUp:
+    address: user/{userId}/signedup
````
Nitpick: for consistency with examples in our reference docs:

```suggestion
    address: 'user/{userId}/signedup'
```

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
````diff
@@ -0,0 +1,84 @@
+      $ref: '#/channels/userSignedUp'  
+```
+
+In this code, the previously defined userId parameter is reused to form the channel name for `UserUpdated` event.
````
Nitpick: for keeping ubiquitous language in place

```suggestion
In this AsyncAPI document, the previously defined userId parameter is reused to form the channel name for `UserUpdated` event.
```

Feel free to simplify to `document` if you prefer.

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
````diff
@@ -0,0 +1,84 @@
+B --> D
+D --> E
+D --> F
+```
````
Will this graph let the user understand better what we are talking about in this doc? I feel this graph is not necessary.

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
```diff
@@ -0,0 +1,84 @@
+    subgraph "Parameter Sources"
+        D --> F[Runtime Event Payload]
+        D --> G[Other possible sources]
+    end
```
I'm not very sure this is providing a clear picture of what we talk about. In fact, I don't think we need a graph here. Same as https://github.com/asyncapi/website/pull/1959/files#r1281588954

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
```diff
@@ -0,0 +1,84 @@
+    parameters:
+      userId:
+        description: Id of the user.
+        location: $message.payload#/userid
```
Nitpick: for consistency with examples in our reference docs:
```suggestion
        location: '$message.payload#/userId'
```

smoya replied:
I would also add the message object to the channel so you can add the `userid` property on it. I believe this will also help users to understand from where the value comes.

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-names.md:
```diff
@@ -0,0 +1,84 @@
+    parameters:
+      userId:
+        description: Id of the user.
+        location: $message.payload#/UserUpdated      
```
This is duplicated. Did you validate this document?

```suggestion
```

mhmohona replied:
I have got the example from here.
![image](https://github.com/asyncapi/website/assets/14244685/56db3c7c-25c2-493e-b8f5-7c4cae15b92b)

However, now I have updated the example as per your other review.

smoya reviewed (changes requested):
I gave my best providing an exhaustive review. However, I did not write all the things I found that need a change.

I feel this document needs a deep rewrite, specially regarding the "tone" of the message we give; it feels very marketing/sales oriented instead of a technical document.

Listing few sentences as example:

- `The Parameter Context clarifies the origin of parameters, ensuring consistent understanding across teams. It also enables efficient code generation, thereby accelerating development and enhancing the developer experience.`
- `are pivotal for enhancing the flexibility, interoperability, reusability, and standardization of asynchronous APIs.`
- ` This adaptability to different APIs aids in creating a unified view of the API, making it easier to comprehend and utilize.`

I will always advocate for short, concise and focused on explain concepts. Skipping fancy words and using just plain English instead.

If you need info about what the feature does, and you feel the current documentation in the [Specification Reference](https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12) is not enough, I'm happy to jump into a call or create a video, etc.

mhmohona committed:
Updated except the file name

I will change the file name later as if I change it now, all reviews which I havent fixed yet, are gonna be outdated.

smoya commented:
@mhmohona you can now easily validate your AsyncAPI v3 docs with https://studio.asyncapi.com. here is how to enable it https://www.loom.com/share/0bab6489e24c4e03b0f7951ed793a240?sid=d98f7310-3d0d-4a4c-ac1f-96e4247d5b74

mhmohona committed:
updating as per review

mhmohona commented:
@smoya thank you for your feedback. I have addressed your review and updated accordingly. Can you please have a look at it again?

quetzalliwrites commented:
@mhmohona **Please review this PR and ensure you resolve EACH feedback comment left.** Sergio left a lot of extremely useful and detailed feedback, but you didn't resolve each item and respond to each one. Let's do that now before we bother Sergio or others for another review because it is hard to tell what feedback you addressed/missed. ✌🏽😄

quetzalliwrites commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
````diff
@@ -0,0 +1,120 @@
+D --> F
+```
+
+The flowchart shows how dynamic channel addresses allow flexible event-driven interfaces by using and reusing parameters in channel names, adapting well to various APIs and protocols.
````
We do not use the word "flowchart" for diagrams in our docs. Let's remember to be consistent, and please update all mentions to say "diagram."

quetzalliwrites commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
```diff
@@ -0,0 +1,120 @@
+
+This document defines a dynamic channel address for a `userSignedUp` event message, making it easy to include and reuse specific details in the channel address.
+
+## Parameter Context
```
We observe sentence capitalization for headers in our docs; please update all headers accordingly. 😄

mhmohona committed:
update as per review

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
```diff
@@ -0,0 +1,125 @@
+weight: 80
+---
+
+Dynamic channel addresses are used to specify dynamic parts of a channel name, which becomes particularly useful when you want to use and reuse parameters in the channel name. It provides flexibility in how you structure your event-driven interfaces.
```
Changing `name` to `address` to be correct. Also simplifying suggestion:

```suggestion
Dynamic channel addresses specify dynamic parts of a channel address, which becomes particularly useful when you want to use and reuse parameters in a channel address.
```

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
````diff
@@ -0,0 +1,125 @@
+      B --> D
+      D --> E
+      D --> F
+```
````
As mentioned in https://github.com/asyncapi/website/pull/1959#discussion_r1281588954, I still believe this diagram doesn't clarify anything to the user but rather can lead to confusion. I would suggest we remove it.

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
````diff
@@ -0,0 +1,125 @@
+Here is an example of dynamic channel address:
+
+```yml
+userSignedUp:
+  address: 'user.signedup'
+  messages:
+    userSignedUp:
+      $ref: '#/components/messages/userSignedUp'
````
This example is not showing any parameter in the address. You are already providing one example below in https://github.com/asyncapi/website/pull/1959/files#diff-798204f3a6dff3c49407fb1af6e8d8d4a10a5b26e328844fe1e03e17173e3c98R66.

mhmohona replied:
So here shall I remove the code example? Or I should provide a new example?

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
```diff
@@ -0,0 +1,125 @@
+    F --> G
+    A --> G
+
+    style B fill:#47BCEE,stroke:#47BCEE;
```
As mentioned in https://github.com/asyncapi/website/pull/1959#discussion_r1281588954, I still believe these diagrams don't clarify anything to the user but rather can lead to confusion. I would suggest we remove it.

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
```diff
@@ -0,0 +1,125 @@
+
+## Parameter context
+
+In a channel address, there's a map of parameters, which needs to include all the same parameters that are in the main channel address. The names you use for the parameters in this map must be exactly the same as the names you used in the channel address.
```
Short and clear 👍 I like it

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
```diff
@@ -0,0 +1,125 @@
+
+This diagram shows how the channel address includes the same parameters as the channel address, and each parameter's name matches the one used in the address.
+
+Here is an example of parameter context:
```
What does `context` mean here? I understand the title above but here doesn't have a meaning IMHO. Suggesting a change:
```suggestion
Here is an example of a dynamic channel address and its parameter:
```

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
````diff
@@ -0,0 +1,125 @@
+      $ref: "#/components/messages/userSignedUp"
+```
+
+In the above document, `user/{userId}/signedup` is the endpoint where `{userId}` is a path parameter.
````
Nitpick: Using ubiquitous language
```suggestion
In the above document, `user/{userId}/signedup` is the address where `{userId}` is a parameter for that address.
```

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
```diff
@@ -0,0 +1,125 @@
+
+## Reusing parameters
+
+Parameters can be reused. This reuse optimizes the usage of parameters within the event-driven interface by letting you use the parameters again.
```
Nitpick: using ubiquitous language
```suggestion
Parameters can be reused. This reuse optimizes the usage of parameters within the API by letting you use the parameters again.
```

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
```diff
@@ -0,0 +1,125 @@
+
+Parameters can be reused. This reuse optimizes the usage of parameters within the event-driven interface by letting you use the parameters again.
+
+If there is another message, for example, a `UserUpdated` message, which also requires the `userId`, the parameter can be reused like the following following diagram:
```
```suggestion
If there is another message, for example, a `UserUpdated` message, which also requires the `userId` parameter, it can be reused like in the following example:
```

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
````diff
@@ -0,0 +1,125 @@
+
+If there is another message, for example, a `UserUpdated` message, which also requires the `userId`, the parameter can be reused like the following following diagram:
+
+```mermaid
+graph TD
+    subgraph Channels
+        userSignedUp["userSignedUp"]
+    end
+
+    subgraph Parameters
+        userId["userId"]
+    end
+
+    userSignedUp -->|requires| userId
+
+    subgraph Reused Parameters
+        UserUpdated["UserUpdated"]
+    style UserUpdated fill:#47BCEE,stroke:#47BCEE;
+    
+    end
+    UserUpdated -->|reuses| userId
+```
+
+In this diagram, a channel named `userSignedUp` that requires a parameter `userId` .Additionally, the parameter `userId` is reused for another message, `UserUpdated`.
````
As mentioned in https://github.com/asyncapi/website/pull/1959#discussion_r1281588954, I still believe these diagrams don't clarify anything to the user but rather can lead to confusion. I would suggest we remove it.

The example below is more than enough to illustrate

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
```diff
@@ -0,0 +1,125 @@
+
+In this diagram, a channel named `userSignedUp` that requires a parameter `userId` .Additionally, the parameter `userId` is reused for another message, `UserUpdated`.
+
+Here is a code of reusing parameters:
```
```suggestion
Here is an example of reusing parameters:
```

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
````diff
@@ -0,0 +1,125 @@
+    action: receive
+    channel: 
+      $ref: '#/channels/userSignedUp'  
+```
````
This example is not reusing any parameter. Also, I'm confused because you are defining `parameters` value twice.

Here I created an example for you:
```yaml
channels:
  UserSignedUp:
    address: 'user/{userId}/signedup'
    parameters:
      $ref: '#/components/parameters/userId'
  UserUpdated:
    address: 'user/{userId}/updated'
    parameters:
      $ref: '#/components/parameters/userId'
components:
  parameters:
    userId:
      description: Id of the user.
```

smoya commented on pages/docs/concepts/asyncapi-document/dynamic-channel-address.md:
````diff
@@ -0,0 +1,125 @@
+      $ref: '#/channels/userSignedUp'  
+```
+
+In this AsyncAPI document, the previously defined `userId` parameter is reused to form the channel address for `UserUpdated` message.
````
Just a suggestion:
```suggestion
In the previous example, the `userId` parameter is defined under `components`, and both channels  `UserSignedUp` and `UserUpdated` are reusing it in their address by referencing via `$ref`.
```

mhmohona committed:
update as per suggetion
